SignIn.jsx

```diff
@@ -0,0 +1,186 @@
+import * as React from "react";
+import Avatar from "@mui/material/Avatar";
+import Button from "@mui/material/Button";
+import CssBaseline from "@mui/material/CssBaseline";
+import TextField from "@mui/material/TextField";
+import FormControlLabel from "@mui/material/FormControlLabel";
+import Checkbox from "@mui/material/Checkbox";
+import Link from "@mui/material/Link";
+import Grid from "@mui/material/Grid";
+import Box from "@mui/material/Box";
+import LockOutlinedIcon from "@mui/icons-material/LockOutlined";
+import Typography from "@mui/material/Typography";
+import Container from "@mui/material/Container";
+import { createTheme, ThemeProvider } from "@mui/material/styles";
+import { useDispatch, useSelector } from "react-redux";
+import {
+  signIn,
+  signOut,
+  selectUser,
+  setPost,
+} from "../redux/features/SignInSlice";
+import {
+  auth,
+  createUserWithEmailAndPassword,
+  updateProfile,
+  signInWithEmailAndPassword,
+  firestore,
+  signInWithGoogle,
+} from "../firebase";
+import { getAuth, signInWithPopup, GoogleAuthProvider } from "firebase/auth";
+import { useNavigate } from "react-router-dom";
+import { useEffect } from "react";
+import { useAuthState } from "react-firebase-hooks/auth";
+import { doc, getDoc, getDocs, collection } from "firebase/firestore";
+
+function Copyright(props) {
+  return (
+    <Typography variant="body2" color="text.secondary" align="center">
+      {"Copyright © "}
+      <Link color="inherit" href="#">
+        Your Website
+      </Link>{" "}
+      {new Date().getFullYear()}
+      {"."}
+    </Typography>
+  );
+}
+export default function Google() {
+  const [user, setuser] = React.useState({
+    firstname: "",
+    lastname: "",
+    username: "",
+    imgurl: "",
+  });
+
+  const theme = createTheme();
+  const dispatch = useDispatch();
+  const curr = useSelector(selectUser);
+  const navigate = useNavigate();
+  const [use, loading, error] = useAuthState(auth);
+  const login = (event) => {
+    event.preventDefault();
+    signInWithEmailAndPassword(auth, user.email, user.password)
+      .then((userAuth) => {})
+      .catch((err) => {
+        alert(err);
+      });
+  };
+  useEffect(() => {
+    if (loading) {
+      return;
+    }
+    if (use) {
+      const uid = use.uid;
+      const getuser = async () => {
```

Comment:
A function doing several tasks can make code reusability and maintainability hard. Consider modularizing the function